sklearn/neighbors/tests/test_ball_tree.py

```diff
@@ -101,3 +104,96 @@ def one_arg_func(x):
+
+    # Test consistency with respect to the `query` method
+    k = 5
+    dist_64, ind_64 = bt_64.query(Y_64, k=k)
+    dist_32, ind_32 = bt_32.query(Y_32, k=k)
+    assert_allclose(dist_64, dist_32, rtol=1e-5)
+    assert_equal(ind_64, ind_32)
+    assert dist_64.dtype == np.float64
+    assert dist_32.dtype == np.float32
+
+    # Test consistency with respect to the `query_radius` method
+    r = 2.38
+    ind_64, neighbors_64 = bt_64.query_radius(Y_64[0:2, :], r=r)
+    ind_32, neighbors_32 = bt_32.query_radius(Y_32[0:2, :], r=r)
+    assert_equal(ind_64, ind_32)
+    assert_allclose(
+        neighbors_64,
+        neighbors_32,
+    )
+
+    # Test consistency with respect to the `query_radius` method
+    # with return distances being true
+    ind_64, dist_64 = bt_64.query_radius(Y_64[4:5, :], r=r, return_distance=True)
+    ind_32, dist_32 = bt_32.query_radius(Y_32[4:5, :], r=r, return_distance=True)
+    assert_equal(ind_64[0], ind_32[0])
+    assert_allclose(dist_64[0], dist_32[0], rtol=1e-5)
+    assert dist_64[0].dtype == np.float64
+    assert dist_32[0].dtype == np.float32
+
+
+@pytest.mark.parametrize("metric", itertools.chain(METRICS, BOOLEAN_METRICS))
+def test_kernel_density_numerical_consistency(global_random_seed, metric):
+    # Test consistency with respect to the `kernel_density` method
+    rng = np.random.RandomState(global_random_seed)
+    _X = rng.random_sample((100, 3))
+    _Y = rng.random_sample((5, 3))
+
+    X_64 = _X.astype(dtype=np.float64, copy=False)
+    Y_64 = _Y.astype(dtype=np.float64, copy=False)
+
+    X_32 = _X.astype(dtype=np.float32, copy=False)
+    Y_32 = _Y.astype(dtype=np.float32, copy=False)
+
+    metric_params = METRICS.get(metric, {})
+    bt_64 = BallTree(X_64, leaf_size=1, metric=metric, **metric_params)
+    bt_32 = BallTree32(X_32, leaf_size=1, metric=metric, **metric_params)
+
+    kernel = "gaussian"
+    h = 0.1
+    density64 = bt_64.kernel_density(Y_64, h=h, kernel=kernel, breadth_first=True)
+    density32 = bt_32.kernel_density(Y_32, h=h, kernel=kernel, breadth_first=True)
+    assert_allclose(density64, density32, rtol=1e-5)
+    assert density64.dtype == np.float64
+    assert density32.dtype == np.float32
+
+
+def test_two_point_correlation_numerical_consistency(global_random_seed):
+    # Test consistency with respect to the `two_point_correlation` method
+    rng = np.random.RandomState(global_random_seed)
+    _X = rng.random_sample((100, 3))
+    _Y = rng.random_sample((5, 3))
+
+    X_64 = _X.astype(dtype=np.float64, copy=False)
+    Y_64 = _Y.astype(dtype=np.float64, copy=False)
+
+    X_32 = _X.astype(dtype=np.float32, copy=False)
+    Y_32 = _Y.astype(dtype=np.float32, copy=False)
+
+    bt_64 = BallTree(X_64, leaf_size=10)
+    bt_32 = BallTree32(X_32, leaf_size=10)
+
+    r = np.linspace(0, 1, 10)
+
+    counts_64 = bt_64.two_point_correlation(Y_64, r=r, dualtree=True)
+    counts_32 = bt_32.two_point_correlation(Y_32, r=r, dualtree=True)
```

Comment:
Since these are of dtype intp_t I think we might not need to check them.
https://github.com/scikit-learn/scikit-learn/blob/01b9050b690ebb07870454720f5f67ac235348e4/sklearn/neighbors/_binary_tree.pxi#L1573-L1589